src/mpl/src/gavl/mpl_gavl.c

```diff
@@ -260,18 +263,72 @@ int MPL_gavl_tree_insert(MPL_gavl_tree_t gavl_tree, const void *addr, uintptr_t
                     gavl_left_rotation(cur_node, rnode);
             }
 
-            if (!STACK_EMPTY(node_stack)) {
-                STACK_POP(node_stack, cur_node);
-                goto stack_recovery;
+            if (!STACK_EMPTY(gavl_tree_iptr->node_stack)) {
+                STACK_POP(gavl_tree_iptr->node_stack, cur_node);
+                continue;
             } else {
                 break;
             }
         } while (1);
 
-        while (gavl_tree_iptr->root->parent != NULL)
+        while (gavl_tree_iptr->root && gavl_tree_iptr->root->parent != NULL)
             gavl_tree_iptr->root = gavl_tree_iptr->root->parent;
     }
 
+    return;
+}
+
+int MPL_gavl_tree_insert(MPL_gavl_tree_t gavl_tree, const void *addr, uintptr_t len,
+                         const void *val)
+{
+    int mpl_err = MPL_SUCCESS;
+    gavl_tree_node_s *node_ptr;
+    gavl_tree_s *gavl_tree_iptr = (gavl_tree_s *) gavl_tree;
+
+    node_ptr = (gavl_tree_node_s *) MPL_malloc(sizeof(gavl_tree_node_s), MPL_MEM_OTHER);
+    if (node_ptr == NULL) {
+        mpl_err = MPL_ERR_SHM_NOMEM;
+        goto fn_fail;
+    }
+    node_ptr->parent = NULL;
+    node_ptr->left = NULL;
+    node_ptr->right = NULL;
+    node_ptr->height = 1;
+    node_ptr->addr = (uintptr_t) addr;
+    node_ptr->len = len;
+    node_ptr->val = val;
```

Comment:
This can use a macro, e.g. `GAVL_NODE_INIT(addr, len, val)`. The init values for internal fields are internal details, hide and separate the context.